### PR TITLE
Adaptive hashmap implementation

### PR DESCRIPTION
**Pull request by arthurprs**
All credits to @pczarn who wrote https://github.com/rust-lang/rfcs/pull/1796 and https://github.com/contain-rs/hashmap2/pull/5

 **Background**

Rust std lib hashmap puts a strong emphasis on security, we did some improvements in https://github.com/rust-lang/rust/pull/37470 but in some very specific cases and for non-default hashers it's still vulnerable (see #36481).

This is a simplified version of https://github.com/rust-lang/rfcs/pull/1796 proposal sans switching hashers on the fly and other things that require an RFC process and further decisions. I think this part has great potential by itself.

**Proposal**
This PR adds code checking for extra long probe and shifts lengths (see code comments and https://github.com/rust-lang/rfcs/pull/1796 for details), when those are encountered the hashmap will grow (even if the capacity limit is not reached yet) _greatly_ attenuating the degenerate performance case.

We need a lower bound on the minimum occupancy that may trigger the early resize, otherwise in extreme cases it's possible to turn the CPU attack into a memory attack. The PR code puts that lower bound at half of the max occupancy (defined by ResizePolicy). This reduces the protection (it could potentially be exploited between 0-50% occupancy) but makes it completely safe.

**Drawbacks**

* May interact badly with poor hashers.  Maps using those may not use the desired capacity.
* It adds 2-3 branches to the common insert path, luckily those are highly predictable and there's room to shave some in future patches.
* May complicate exposure of ResizePolicy in the future as the constants are a function of the fill factor.

**Example**

Example code that exploit the exposure of iteration order and weak hasher.

```
const MERGE: usize = 10_000usize;
#[bench]
fn merge_dos(b: &mut Bencher) {
    let first_map: $hashmap<usize, usize, FnvBuilder> = (0..MERGE).map(|i| (i, i)).collect();
    let second_map: $hashmap<usize, usize, FnvBuilder> = (MERGE..MERGE * 2).map(|i| (i, i)).collect();
    b.iter(|| {
        let mut merged = first_map.clone();
        for (&k, &v) in &second_map {
            merged.insert(k, v);
        }
        ::test::black_box(merged);
    });
}
```

_91 is stdlib and _ad is patched (the end capacity in both cases is the same)

```
running 2 tests
test _91::merge_dos              ... bench:  47,311,843 ns/iter (+/- 2,040,302)
test _ad::merge_dos              ... bench:     599,099 ns/iter (+/- 83,270)
```

### CODE REVIEW

**Comment by arthurprs** [comment]
The code is in very rough shape, I wanted to collect feedback on the idea first.

**Comment by alexcrichton** [comment]
r? @bluss 

cc @pczarn, @apasel422

File: src/libstd/collections/hash/map.rs [hunk 1]
```diff
@@ -708,10 +718,11 @@ impl<K, V, S> HashMap<K, V, S>
         assert!(self.table.size() <= new_raw_cap);
         assert!(new_raw_cap.is_power_of_two() || new_raw_cap == 0);
 
+        self.long_probes = false;
         let mut old_table = replace(&mut self.table, RawTable::new(new_raw_cap));
         let old_size = old_table.size();
 
-        if old_table.capacity() == 0 || old_table.size() == 0 {
+        if old_table.size() == 0 {
```
Comment (bluss) [inline src/libstd/collections/hash/map.rs]: why was the capacity conditional removed here?

**Comment by arthurprs** [comment]
I can remove that change. But the capacity check is redundant, right?

File: src/libstd/collections/hash/map.rs [hunk 2]
```diff
@@ -1962,8 +1978,20 @@ impl<'a, K: 'a, V: 'a> VacantEntry<'a, K, V> {
     #[stable(feature = "rust1", since = "1.0.0")]
     pub fn insert(self, value: V) -> &'a mut V {
         match self.elem {
-            NeqElem(bucket, disp) => robin_hood(bucket, disp, self.hash, self.key, value),
-            NoElem(bucket) => bucket.put(self.hash, self.key, value).into_mut_refs().1,
+            NeqElem(bucket, disp) => {
+                let (shift, v_ref) = robin_hood(bucket, disp, self.hash, self.key, value);
+                if disp >= 128 || shift >= 512 {
```
Comment (arthurprs) [inline src/libstd/collections/hash/map.rs]: These of course will be moved into well commented constants.

**Comment by bluss** [comment]
Looks remarkably simple for what it does. That's good.

Advantages:

+ Only affects insert (the fast lookup we have is untouched)
+ Simple

Obviously the constants involved need constant names, tuning and comments. I think we can make an argument that for example a displacement of 128 slots from its best position is always a bad case and should *never* occur in a healthy hash table, no matter its size?

File: src/libstd/collections/hash/map.rs (hunk 1 above)
Comment (arthurprs) [inline src/libstd/collections/hash/map.rs]: This doesn't need to be part of the PR. The capacity check is redundant though, right?

File: src/libstd/collections/hash/map.rs (hunk 2 above)
Comment (arthurprs) [inline src/libstd/collections/hash/map.rs]: I wonder if we can get away only checking probe length. It's still possible to abuse long shifts without hitting the probe length limit but that's a lot harder.

**Comment by arthurprs** [comment]
The good thing is that the math behind this is independent from the map size, it's only a function of the fill factor and the hasher being good. The first is fine as the constants work for fill factors smaller than the one they were calculated.

Interacting badly with bad hashers could be problematic in practice as hashmap may never reach the maximum fill factor (the check for half filled is useful here so it doesn't blow up).

**Comment by Veedrac** [comment]
Although this makes it harder to trigger pathological behaviour, the risk here is that if you *do* manage, the damage can be exponential rather than just quadratic.

I'll look over this in more detail over the next few days, but the most immediate risk is that the hash map might become literally unusable for bad hashing schemes like identity hashes, and extremely risky for ones where collisions are known.

**Comment by Veedrac** [comment]
> We need a lower bound on the minimum occupancy that may trigger the early resize, otherwise in extreme cases it's possible to turn the CPU attack into a memory attack. The PR code puts that lower bound at half of the max occupancy (ResizePolicy).

This sounds like a good idea, but it means it only counters the n=2 case. That's definitely an improvement, though.

**Comment by arthurprs** [comment]
@Veedrac what do you mean by n=2 case?

**Comment by arthurprs** [comment]
Putting into more generic terms, you mean that it can still be abused while between 0% and 50% filled?

**Comment by Veedrac** [comment]
Yes, basically. I'll try to cook up some examples later, to give a more concrete demo.

**Comment by arthurprs** [comment]
_Trying to resume the conversation..._ I think the obvious open question here is the iteration with less than good hashers, hashmaps using those may not use the desired capacity.

**Comment by alexcrichton** [comment]
The libs team discussed this briefly at triage the other day, and we were wondering if we could perhaps land this ahead of the RFC? The changes to probing here are universally better, even if we don't do the hasher changes yet, right?

If so perhaps, the PR title/description could be cleaned up to the current state and we could look to merge?

**Comment by arthurprs** [comment]
I'll update the PR/description to provide clearer picture.

> The changes to probing here are universally better, even if we don't do the hasher changes yet, right?

I wouldn't say universally but mostly.

**Comment by alexcrichton** [comment]
Ah ok, thanks for the clarification. Want to ping me when updated and we can look to merge?

**Comment by arthurprs** [comment]
I should have elaborated that. It's not strictly better because the interaction with poor hashers isn't great, with those it's possible that the hashmap resizes early even on non-rogue input.

I'll ping when I update it.

**Comment by arthurprs** [comment]
PR updated, now there's two constants and lots of comment lines.

**Comment by alexcrichton** [comment]
Thanks @arthurprs! Out of curiosity, would it be at all possible to add a test for this?

**Comment by arthurprs** [comment]
I think so. It's possible to observe the early resizes from the public api and it's somewhat easy to trigger it by merging two maps with the same hash seed (like the example in first post). I'll write something tomorrow.

**Comment by alexcrichton** [comment]
@bors: r+

Thanks!

**Comment by alexcrichton** [comment]
Oh wait looks like a merge conflict now :(

**Comment by arthurprs** [comment]
I rebased and squashed the commits.

**Comment by alexcrichton** [comment]
@bors: r+

Thanks again and for being patient @arthurprs!

**Comment by arthurprs** [comment]
@istankovic Please make a PR :smiley:

**Comment by istankovic** [comment]
@arthurprs Nah, it was just something I noticed so I made the comments, but it doesn't bother me enough to make a PR, sorry...

**Comment by arthurprs** [comment]
The shift length math is broken. It turns out that checking for the shift length is complicated. Using simulations it's possible to see that a value of 2000 will only get probabilities down to ~1e-7 when the hashmap load factor is 90% (rust goes up to 90.9% as of today). That's probably not good enough to go into the stdlib with pluggable hashers. See https://github.com/rust-lang/rfcs/pull/1796#issuecomment-280931018

I suggest taking that part out and keeping only displacement check, which is much safer and very useful by itself.

cc @pczarn @bluss @alexcrichton

**Comment by pczarn** [comment]
I agree. This issue also indicates that the hashmap load factor may be too high.

Thanks for help running these simulations.

**Comment by SimonSapin** [comment]
Because of alignment(?), this one `bool` field makes the already-large `HashMap` grow from 40 bytes to 48. Maybe it’s worth the space, but I thought this should be noted.

(We caught this in Servo because we have unit tests that check `std::mem::size_of` for various types used in DOM nodes, to catch size regressions. Because there can be so many nodes in a document, even a small size increase can be significant.)

**Comment by SimonSapin** [comment]
Could the extra bit be packed in `RawTable::size` or `RawTable::capacity`?

**Comment by pczarn** [comment]
Yes, of course.

If we're able to restrict adaptive hashing to maps with the default hasher, I'd prefer to have the extra bit in `SipHasher`.

**Comment by arthurprs** [comment]
It's just a matter of finding how to use the bit with reasonable code.

I'd argue against making it RandomState only, the selling point was supporting all hashmaps.

**Comment by SimonSapin** [comment]
Let’s discuss in #40042.

**Pull request merged**